R/container.R

```diff
@@ -0,0 +1,174 @@
+#' Declare post-processing for model predictions
+#'
+#' @param mode The model's mode, one of `"unknown"`, `"classification"`, or
+#' `"regression"`. Modes of `"censored regression"` are not currently supported.
+#' @param type The model sub-type. Possible values are `"unknown"`, `"regression"`,
+#' `"binary"`, or `"multiclass"`.
+#' @param outcome The name of the outcome variable.
+#' @param estimate The name of the point estimate (e.g. predicted class)
+#' @param probabilities The names of class probability estimates (if any). For
+#' classification, these should be given in the order of the factor levels of
+#' the `estimate`.
+#' @param time The name of the predicted event time. (not yet supported)
+#' @examples
+#'
+#' container()
+#' @export
+container <- function(mode = "unknown", type = "unknown", outcome = character(0),
+                      estimate = character(0), probabilities = character(0),
+                      time = character(0)) {
+  dat <-
+    list(
+      outcome = outcome,
+      type = type,
+      estimate = estimate,
+      probabilities = probabilities,
+      time = time
+    )
+
+  new_container(
+    mode,
+    type,
+    operations = list(),
+    columns = dat,
+    ptype = tibble::new_tibble(list()),
+    call = current_env()
+  )
+}
+
+new_container <- function(mode, type, operations, columns, ptype, call) {
+  mode <- arg_match0(mode, c("unknown", "regression", "classification", "censored regression"))
+
+  if (mode == "regression") {
+    type <- "regression"
+  }
+
+  type <- arg_match0(type, c("unknown", "regression", "binary", "multiclass"))
+
+  if (!is.list(operations)) {
+    cli_abort("The {.arg operations} argument should be a list.", call = call)
+  }
+
+  is_oper <- purrr::map_lgl(operations, ~ inherits(.x, "operation"))
+  if (length(is_oper) > 0 & !any(is_oper)) {
+    bad_oper <- names(is_oper)[!is_oper]
+    cli_abort("The following {.arg operations} do not have the class \\
+                   {.val operation}: {bad_oper}.", call = call)
+  }
+
+  # validate operation order and check duplicates
+  validate_order(operations, mode, call)
+
+  # check columns
+  res <- list(
+    mode = mode, type = type, operations = operations,
+    columns = columns, ptype = ptype
+  )
+  class(res) <- "container"
+  res
+}
+
+#' @export
+print.container <- function(x, ...) {
+  # todo emulate Emil's recipe printing
+
+  num_op <- length(x$operations)
+  cli_inform("{x$type} post-processing object with {num_op} operation{?s}")
+
+  if (num_op > 0) {
+    cat("\n")
+    res <- purrr::map(x$operations, ~ print(.x))
+  }
+
+  invisible(x)
+}
+
+#' @export
+fit.container <- function(object, .data, outcome, estimate, probabilities = c(),
+                          time = c(), ...) {
+  # ------------------------------------------------------------------------------
+  # set columns via tidyselect
+
+  dat <- list()
```

Comment:
`dat` is somewhat confusing when a `.data` lives in the same environment. Maybe `columns`, as it's eventually named in the object?

Reply:
👍 [ab07580](https://github.com/tidymodels/container/pull/1/commits/ab075802f32dc66969c8069fd5efb1e3afe53315)